draft-ietf-tls-tls13.md

```diff
@@ -1914,19 +1916,23 @@ Structure of this message:
 %%% Key Exchange Messages
 
        struct {
-           ProtocolVersion version;
+           ProtocolVersion legacy_version = 0x0303;    /* TLS v1.2 */
            Random random;
+           opaque legacy_session_id_echo<0..32>;
            CipherSuite cipher_suite;
+           uint8 legacy_compression_method = 0;
            Extension extensions<6..2^16-1>;
        } ServerHello;
 
 version
```

Comment:
nit: legacy_version

Reply:
Yea, I don't really like it either to have two version fields.

Reply:
Me neither, but it's that or an insecure version fallback. Aesthetics and cleanliness are valuable, but not at the cost of security.

Reply:
But the same problem may repeat with future middleboxes.

Reply:
Future middleboxes should understand TLSv1.3... If they don't then it's the manufacturers fault.

Reply:
Indeed. For the future, we need to better take into account how to defend our extension points when we design protocols. There's things like GREASE, but GREASE doesn't go far enough here. I have various ideas to try out on this front.

In the meantime, because of how much we've let ossify up to now, we sadly must live with certain constraints if we want TLS 1.3 to simultaneously be deployable and secure.

Reply:
Is this the most common problem though?

Reply:
Common enough. I'm surprised you're bothered by this particular piece. We've already moved the ClientHello version out.

Reply:
I am talking a list with only one entry though, and you are talking about middleboxes not servers too.

Reply:
We have a few local middleboxes to test with. The setup is nowhere near exhaustive (the variance on these things is *huge*), but it did allow us to iterate on this locally with a number of designs before testing a few out into the wild. This PR came out of what we've learned from all those experiments.

Variants that didn't fix the ServerHello version didn't even survive the local pass. :-(

Reply:
So do you know why these individual middleboxes did it?

Reply:
Also consider that most know or can find out if they are using middleboxes or not, so an "insecure version fallback" is not the only alternative.

Reply:
Mostly the only way to know is to measure, and that's just a network attacker.

Reply:
Is getting into an arms race with middleboxes really a good idea?